api/v2/server.go

```diff
@@ -0,0 +1,183 @@
+package v2
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/cockroachdb/cmux"
+	"github.com/grpc-ecosystem/go-grpc-prometheus"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	pb "github.com/coreos/clair/api/v2/clairpb"
+	"github.com/coreos/clair/database"
+)
+
+func errorHandle(err error) {
+	if err != nil {
+		if opErr, ok := err.(*net.OpError); !ok || (ok && opErr.Op != "accept") {
+			log.Fatal(err)
+		}
+	}
+}
+
+var (
+	promResponseDurationMilliseconds = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "clair_v2_api_response_duration_milliseconds",
+		Help:    "The duration of time it takes to receieve and write a response to an V2 API request",
+		Buckets: prometheus.ExponentialBuckets(9.375, 2, 10),
+	}, []string{"route", "code"})
+)
+
+func init() {
+	prometheus.MustRegister(promResponseDurationMilliseconds)
+}
+
+func newGrpcServer(paginationKey string, store database.Datastore, tlsConfig *tls.Config) *grpc.Server {
+	grpcOpts := []grpc.ServerOption{
+		grpc.UnaryInterceptor(grpc_prometheus.UnaryServerInterceptor),
+		grpc.StreamInterceptor(grpc_prometheus.StreamServerInterceptor),
+	}
+	if tlsConfig != nil {
+		grpcOpts = append(grpcOpts, grpc.Creds(credentials.NewTLS(tlsConfig)))
+	}
+	grpcServer := grpc.NewServer(grpcOpts...)
+	pb.RegisterClairAPIServiceServer(grpcServer, &ClairAPIServer{PaginationKey: paginationKey, Store: store})
+	return grpcServer
+}
+
+type httpStatusWritter struct {
+	http.ResponseWriter
+
+	StatusCode int
+}
+
+func (w *httpStatusWritter) WriteHeader(code int) {
+	w.StatusCode = code
+	w.ResponseWriter.WriteHeader(code)
+}
+
+func logHandler(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		start := time.Now()
+		lrw := &httpStatusWritter{ResponseWriter: w, StatusCode: http.StatusOK}
+		handler.ServeHTTP(lrw, r)
+		statusStr := strconv.Itoa(lrw.StatusCode)
+		if lrw.StatusCode == 0 {
+			statusStr = "???"
+		}
```

Comment:
We should probably stop doing this and and just assume all Prometheus requests responded with 200, but that can be handled in a separate PR.